lib/msal-core/src/telemetry/ApiEvent.ts

```diff
@@ -0,0 +1,92 @@
+import TelemetryEvent from "./TelemetryEvent";
+import { EVENT_NAME_PREFIX, TELEMETRY_BLOB_EVENT_NAMES } from "./TelemetryConstants";
+import { scrubTenantFromUri, hashPersonalIdentifier } from "./TelemetryUtils";
+import { Logger } from "../Logger";
+
+// export these to test... could namespace them, or find another way?
+export const AUTHORITY_KEY = EVENT_NAME_PREFIX + "authority";
+export const AUTHORITY_TYPE_KEY = EVENT_NAME_PREFIX + "authority_type";
+export const PROMPT_KEY = EVENT_NAME_PREFIX + "ui_behavior";
+export const TENANT_ID_KEY = EVENT_NAME_PREFIX + "tenant_id";
+export const USER_ID_KEY = EVENT_NAME_PREFIX + "user_id";
+export const WAS_SUCESSFUL_KEY = EVENT_NAME_PREFIX + "was_successful";
+export const API_ERROR_CODE_KEY = EVENT_NAME_PREFIX + "api_error_code";
+export const LOGIN_HINT_KEY = EVENT_NAME_PREFIX + "login_hint";
+
+export const API_CODE = {
+    AcquireTokenRedirect: 2001,
+    AcquireTokenSilent: 2002,
+    AcquireTokenPopup: 2003,
+    LoginRedirect: 2004,
+    LoginPopup: 2005
+};
+
+export const API_EVENT_IDENTIFIER = {
+    AcquireTokenRedirect: "AcquireTokenRedirect",
+    AcquireTokenSilent: "AcquireTokenSilent",
+    AcquireTokenPopup: "AcquireTokenPopup",
+    LoginRedirect: "LoginRedirect",
+    LoginPopup: "LoginPopup"
+};
+
+
+export default class ApiEvent extends TelemetryEvent {
+
+    private logger: Logger;
+
+    constructor(correlationId: string, logger: Logger) {
+        super(`${EVENT_NAME_PREFIX}api_event`, correlationId);
+        this.logger = logger;
+    }
+
+    public set apiEventIdentifier(apiEventIdentifier: string) {
+        this.event[TELEMETRY_BLOB_EVENT_NAMES.ApiTelemIdConstStrKey] = apiEventIdentifier;
+    }
+
+    public set apiCode(apiCode: number) {
+        this.event[TELEMETRY_BLOB_EVENT_NAMES.ApiIdConstStrKey] = apiCode;
+    }
```

Comment:
General comment: if you `return this;` in all of these setters it allows for chaining.

Reply:
I would understand teh `set` function if you had matching `get` functions, but since you don't, I don't think a lot of value is added by having them be `set` functions.

Reply:
so the usage i see is something like this 

```js
const event = new HttpEvent(params);
TelemetryManager.startEvent(Event);

event.url = url;

event.
fetch(url).then(res => {
    event.httpResponse = 200;
})
.catch(err  => {
    event.errorCode = "FAIL"
    event.httpResponse = 500;
})

TelemetryManager.endEvent(event);

TelemetryManager.flush();


```


I think set is kinda shorthand for setter functions, no?

Reply:
I believe it would be the difference between that and this: 

```js
event.setErrorCode("FAIL").setHttpResponse(500);
```

Not a big deal, ultimately.

Reply:
Same comment as before:

I think it is better to follow JS Best practices here. Implementing chaining may mess up inheritance and in general doesn't follow command-query separation principle, so I would suggest using setters.

Reply:
internal implementation developers never touch, so if we decide to change internally we can. leaving for now